Update flatten to accept DataFrame and GeoDataFrame

FelipeSBarros opened the pull request:
@cuducos the purpose of this PR is to follow the development of `flatten` function (#96), making it accept and work with `DataFrame`;

FelipeSBarros commented on tests/test_flatten.py:
```diff
@@ -8,6 +11,9 @@
     }
 ]
+PD_DATA = DataFrame(DICT_DATA)
+GEOMETRY = [Point(4, 2)]
```
Althou the idea is to start working with DataFram, I have already added those constants, so after finishing working with DataFrame, I could add GeoDataFrame support also.

cuducos replied:
So should you update the PR description saying this PR focus on both (dataframe and geodataframe)?

FelipeSBarros replied:
indeed.

FelipeSBarros commented on tests/test_flatten.py:
```diff
@@ -31,3 +37,21 @@ def test_flatten_dict():
+    flattened_pd = flatten(
+        PD_DATA, nested_columns=["contextInfo", "neighborhood"]
+    )
+    assert flattened_pd.equal(
```
Not sure here, but I understand that to validate if tow DataFrames are equals, we should use [`pd1.equal(pd2)`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.equals.html)

FelipeSBarros replied:
BUT the [test is failing](https://github.com/FelipeSBarros/crossfire/actions/runs/8010990526/job/21883232212?pr=103#step:6:46) in the [data validation](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/crossfire/__init__.py#L58):
```python
def flatten(data, nested_columns=None):
    nested_columns = set(nested_columns or NESTED_COLUMNS)
    if not nested_columns.issubset(NESTED_COLUMNS):
        raise NestedColumnError(nested_columns)
    if not data:
        return data
```
This is because, when dealing with `DataFrame` we need to use different approaches, like `df1.notnull()`, in instance to validate if a DataFram is passed bu it is None...
Thought about passing this validation to underneth of data type validation, and write a new validation to "identify" `DataFrame`s.
Something like:
```python
def flatten(data, nested_columns=None):
    nested_columns = set(nested_columns or NESTED_COLUMNS)
    if not nested_columns.issubset(NESTED_COLUMNS):
        raise NestedColumnError(nested_columns)
    if isinstance(data, list):
        if not data:
            return data
        keys = set(data[0].keys()) & nested_columns
    if isinstance(data, DataFrame):
        if data.isnull():
            return data

```

Is it a messy approach?

cuducos replied:
If the tests are failing you are still on step 2. Asking _is it too messy_ is step 3.

By now focus on making the test pass.

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -6,6 +6,12 @@
 from crossfire.clients import AsyncClient, Client  # noqa
 from crossfire.errors import NestedColumnError
 
+try:
+    from pandas import DataFrame
+except ImportError:
+    pass
+
```
Not sure if I need to use this way... I was imagening in case a user dind't have installed Pandas...

cuducos replied:
If you run it without having Pandas and it does not break, it is OK by now : )

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -52,15 +58,19 @@ def occurrences(
 
 
 def flatten(data, nested_columns=None):
+    origin_type = type(data)
```
Getting this info to use latter, mainly for  `DataFrame` as, for now, I convert to `dict` to flatten and later I will need to return in the same class as it was in the input...

cuducos replied:
Why gatting it now if you might not _use_ it later (e.g. wrong value in `nested_column`)?
Also, why use `isinstance` in some places and compare a type to `origin_type` in other places?

FelipeSBarros replied:
> Why gatting it now if you might not _use_ it later (e.g. wrong value in `nested_column`)?

Make sense. I will put after  the `nested_columns` validation, but before the class validation.

> Also, why use isinstance in some places and compare a type to origin_type in other places?

I can change to use only the `origin_type` in those kind of validation. I woun't be able to use `isinstance`, because I might convert pandas to dict.... (I will pretend I didn't read yet the next mensage. ;)

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -52,15 +58,19 @@ def occurrences(
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    if isinstance(data, DataFrame):
+        data = data.to_dict(orient="records")
```
So, considering your [provocation](https://github.com/FelipeSBarros/crossfire/issues/96#issuecomment-1960282545) I thought it would be easier to make the test pass by converting `DataFrame` to `dcit`, to run the flatten process....

cuducos replied:
I think it is too expensive to convert a DataFrame to dict to just convert it back to DataFrame later… my provocation showed you a way to deal with different logic depending on the type of the input without using `type(…)` or `isinstance(…)`. I am not sure what you meant here…

FelipeSBarros replied:
> I think it is too expensive to convert a DataFrame to dict to just convert it back to DataFrame later… 

I could imagine that... but my goal was to make the test pass... :upside_down_face: 

It is evident that I dind't get your provocation, right? I understand that, you can build a function `add` accepting both, string and integers, by casting (not sure if this is the correct concept/vocalulary) whatever comes to `int`. So, like you said, with no need of data type validation.
But I can't see were/how to apply the same logic in `flatten` function. For now I should expect `dict` or `DataFrame`. But there is no way to `cast` (?) from one to another in a straight way as ``DataFrame` to be converted to a `dict` I must use the method `to_dict` which doesn exist for `dict`....
Other possibility is to convert what comes to `DataFrame`, but in this case, the tests for `dict` will fail.

I am going to far?

cuducos replied:
My provocation had two functions with exactly the same name (`add`). One deals with arguments that are integer numbers. The other deals with arguments that are strings. 

Now replace in that sentence:
* `add` by `flatten`
* _integer numbers_ by _dictionaries_
* _strings_ by _Pandas DataFrame_

FelipeSBarros replied:
> My provocation had two functions with exactly the same name (add). 

Is that possible? To have two functions with the same name?
... goin to check. I will be right back

cuducos replied:
> Is that possible?

[Copy, paste and run it](https://github.com/FelipeSBarros/crossfire/issues/96#issuecomment-1960282545). Then you tell me : )

FelipeSBarros replied:
Sorry, @cuducos I think I dind't get your point...

cuducos replied:
https://github.com/FelipeSBarros/crossfire/assets/4732915/ce4d49b5-2bf1-4f59-bba0-e6d5fe1def3c

FelipeSBarros replied:
Sorry I did what you show in the video, but had to left without time to write about.
After your comment I thought that it as posible to have two function with the same name but different evaluation.
So I try it ad, as I could see/understand the second function overwrite the first one.

Also I didn't get your point with the provocation. Some how your provocation called my attention to the idea of  "casting" the `data` inpiut to be flatten. This is why I did this tranformation: If `flatten` function already works with `dict`, to make the test pass once receving a `DataFrame`, is just a matter of transforming it to `dict`.
Of course, I believe that this would be the soluction fornow...
But now I am completely lost. ...And I am sorry about it.

cuducos replied:
> So I try it and, as I could see/understand, the second function overwrite the first one.
> So I discarted this suposition...

`git commit` and `git push` or didn't happen.

FelipeSBarros replied:
> `git commit` and `git push` or didn't happen.

Ah, I just tested the `add` function to "see" what happens, just like you did in the video.
If I understand right I should duplicate `flatten` function, the first to recieve `dict` and the second, to recieve `DataFrame`.
Right?

FelipeSBarros replied:
@cuducos after much thinking, I decided to try again the test with your provocation. This time I recorded, got the courage to use nvim and added a print to identify the add function. But now I am more confused than before. I can see that this first function ocupied a different memory space then the second. What I don't know is wheter both still existing, and is being used according to the input...
 

https://github.com/FelipeSBarros/crossfire/assets/7208674/5d3c662e-eb15-41ae-ad4e-957708f07757

FelipeSBarros replied:
Another test I did: I changed the restun adding a f-string to confirm if both functions still existing and being used according to the data input.
I can see that indeed the second function overwriten the first one, ocupying a noto memmory space. So event when calling `add` function with `interger` the last function in used, converting it to `int` event not being necessary.

https://github.com/FelipeSBarros/crossfire/assets/7208674/7f948693-1a6b-4faf-8e7e-a4fe26ceef10

FelipeSBarros replied:
Reading back yout [commen](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501166414)


> My provocation had two functions with exactly the same name (`add`). One deals with arguments that are integer numbers. The other deals with arguments that are strings.
> 
> Now replace in that sentence:
> 
> * `add` by `flatten`
> * _integer numbers_ by _dictionaries_
> * _strings_ by _Pandas DataFrame_

Makes me think exactly what I dad before, and have mentioned [here](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501293832)

> Also, I didn't get your point with the provocation. Some how your provocation called my attention to the idea of "casting" the `data` input to be flatten. This is why I did this tranformation: If `flatten` function already works with `dict`, to make the test pass once receving a `DataFrame`, is just a matter of transforming it to `dict`.
> Of course, I believe that this would be the solution for now, to make the test pass and then to start improving...

If converting between `dict` and `DataFrame` was not your point, what was it?

cuducos replied:
Can you commit and push _or_ share your code so I don't have to copy it from a video if I wanna reproduce, comment and tweak it?

cuducos replied:
Ow! I just read that this experimental feature is removed and only used for type-checking! Oh my goddess, sorry I made you spend so much time on this dead-end! Totally my fault!

FelipeSBarros replied:
Don't worry. It is a relieve to realize that I am not crazy.
But, just to confirm: the *experimental feature* you mention is the possibility of having two functions with different parameters/processes?

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -52,15 +58,19 @@ def occurrences(
+        for key in keys:
+            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
+            item.pop(key)
+    if origin_type == DataFrame:
+        return DataFrame(data)
```
To then, retun te result as `DataFrame`...

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -52,15 +58,19 @@ def occurrences(
+        for key in keys:
+            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
+            item.pop(key)
+    if origin_type == DataFrame:
```
Ok. I can't validate with `isinstance` here, becouse I already have change de object type.
I have restructured and commited new proposal.
I will comment on codereview

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -57,20 +57,24 @@ def occurrences(
     )
 
 
+def _make_flatten(data, keys):
```
I have separated this parto considering the idea of DRY.
and, ofcourse, considering that, for now, we will keep converting `DataFrame` to `dict`.

cuducos replied:
I usually think that prefixes like `get,` `load,` and `make` hardly ever contribute to the semantics of the code. If you call a function, it might be implied that it is actioning (getting something, making something, loading something). Thus, if `flatten` is the primary function and you want part of it to be more _internal_, I would use `_flatten` for the “auxiliary” function.

FelipeSBarros replied:
Good point. I have changed this auxiliary function back to `flatten` function and created anotherone for `DataFrame`, called `flatten_df`. I will change it to `_flatten_df`,  now.

FelipeSBarros reviewed (commented):
@cuducos what do you think about this proposal?

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +57,24 @@ def occurrences(
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    if isinstance(data, DataFrame) and not data.empty:
```
What is `DataFrame` here if we end up in the `pass` from line 12?

cuducos replied:
Why don't you check [pur `parser.py`](https://github.com/FelipeSBarros/crossfire/blob/771a750a96de9398d28b44790618b1d85ed0f41c/crossfire/parser.py)?

FelipeSBarros replied:
Actually, yes... the try except came from there. But now I can see that I am not considering the "flag" `HAS_PANDAS`.
I will consider using it in the following commit.

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +57,24 @@ def occurrences(
-                item.pop(key)
+    if isinstance(data, DataFrame) and not data.empty:
+        keys = set(data.columns) & nested_columns
+        data = data.to_dict(orient="records")
```
> I already have change de object type.

The change I commented on didn't make a lot of sense. Why not get rid of it instead? So:
* better performance (less needed transformation)
* more data type safety (Pandas DataFrame types differ from Python dictionaries)
* Avoid something that Python's official doc recommends avoiding

FelipeSBarros replied:
> better performance (less needed transformation)
I am about to commit a change I ahve done. But somehow I still using dict...

 > more data type safety (Pandas DataFrame types differ from Python dictionaries)
How could I improve this point?

> Avoid something that Python's official doc recommends avoiding
What exacltly I should avoid? using data type to chenck object class?

cuducos replied:
> How could I improve this point?

Ifd the input is a DataFrame _and_ the output is also a DataFrame, do not convert it to a dictionary in the middle of the function.

> What exacltly I should avoid?

[Using `type(blablabla) == DataFrame`](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501697218).

FelipeSBarros replied:
Ok. I am not using it anymore.

FelipeSBarros reviewed (commented):
Not sure if I should be commiting this, but as you always says: it is better to discuss considering code implemented  then ideas (actualy you never said exactly that. but always ask to see the code... so, this is how I understand your claim.

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -57,12 +58,11 @@ def occurrences(
-            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-            item.pop(key)
-    return data
+def flatten_df(row, column_name):
+    column_data = row[column_name]
+    return Series(
+        {f"{column_name}_{key}": value for key, value in column_data.items()}
+    )
```
I had to use dict here to set the column name as we are doing with dict `contextInfo_context1`...

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -57,12 +58,11 @@ def occurrences(
-    for item in data:
-        for key in keys:
-            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-            item.pop(key)
```
As I am not converting anymore to dict, I put this code back in the `flatten` function when dealing with dict...

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -71,10 +71,21 @@ def flatten(data, nested_columns=None):
         raise NestedColumnError(nested_columns)
     if isinstance(data, DataFrame) and not data.empty:
         keys = set(data.columns) & nested_columns
-        data = data.to_dict(orient="records")
```
No converting to dict, anymore....

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -71,10 +71,21 @@ def flatten(data, nested_columns=None):
         return data
     keys = set(data[0].keys()) & nested_columns
-    data = _make_flatten(data, keys)
+    for item in data:
+        for key in keys:
+            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
+            item.pop(key)
```
The old `_make_flatten` function.

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -71,10 +71,21 @@ def flatten(data, nested_columns=None):
+                axis=1,
+            )
+
+        return data
```
Working with `DataFrame`.

cuducos commented on crossfire/__init__.py:
```diff
@@ -57,20 +57,24 @@ def occurrences(
+    if isinstance(data, DataFrame) and not data.empty:
+        keys = set(data.columns) & nested_columns
         data = data.to_dict(orient="records")
+        return DataFrame(_make_flatten(data, keys))
```
> converting `DAtaFrame` to `dict`

I still think this [is not a good strategy](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501812069), and you have not tried other alternatives. Let's focus on an alternative to it instead of advancing with this (arguably) unneeded transformation.

cuducos commented on crossfire/__init__.py:
```diff
@@ -6,6 +6,13 @@
 
+try:
+    from pandas import DataFrame, Series
+    from pandas import concat as pd_concat
```
Why `pd_concat` instead of just `concat`?

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +58,34 @@ def occurrences(
+        keys = set(data.columns) & nested_columns
+        for key in keys:
+            data = pd_concat(
+                [
+                    data.drop(key, axis=1),
+                    data.apply(flatten_df, args=(key,), axis=1),
+                ],
```
Not really important, but [no need to use a list if we're not mutating the list](https://cuducos.me/2018/06/13/tuplas-e-listas-no-python.html):

```suggestion
                (
                    data.drop(key, axis=1),
                    data.apply(flatten_df, args=(key,), axis=1),
                ),
```

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +58,34 @@ def occurrences(
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    elif not data:
         return data
```
This should be an _early return_, not a late one.

cuducos replied:
If you pass and empty `DataFrame`, what happens?

There is a bug there and an opportunity for optimization you are missing. 

An early return means returning before engaging in more demanding computing power, I/O or other resources. Before transforming stuff, you can check if it's an empty sequence (e.g., list) or an empty `DataFrame`.

cuducos reviewed (changes requested):
Have you tested without `pandas` installed?

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -8,8 +8,10 @@
+    HAS_PANDAS = True
 except ImportError:
-    pass
+    HAS_PANDAS = False
```
Added to consider in `data` input validation...

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -68,19 +70,23 @@ def flatten(data, nested_columns=None):
-
-        return data
-    elif not data:
+    if HAS_PANDAS and isinstance(data, DataFrame):
+        if data.empty:
+            return data
```
Tried to improve the early return in case of `DataFrame`...

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -68,19 +70,23 @@ def flatten(data, nested_columns=None):
+
+            return data
+
+    if not data:
         return data
```
I know, still not being an early return... but I coun't figure out a way to put this in the beggining with repeating some validations...

cuducos replied:
You can use the `HAS_PANDAS` to know whether to check for an empty list or an empty `DataFrame`, can't you?

FelipeSBarros replied:
Yes. I believe that i what I am doing [here](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1503256591). Right?

FelipeSBarros reviewed (commented):
I know that this last commit doesn't solve the concerns your mentioned before, about the early return. But tried to work improve in caso the `data` is `DAtaFrame`....

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -70,6 +70,9 @@ def flatten(data, nested_columns=None):
     nested_columns = set(nested_columns or NESTED_COLUMNS)
     if not nested_columns.issubset(NESTED_COLUMNS):
         raise NestedColumnError(nested_columns)
+    if not HAS_PANDAS or (HAS_PANDAS and not isinstance(data, DataFrame)):
+        if not data:
+            return data
```
What about this?

FelipeSBarros replied:
Because the user can have pandas but for a specific reason the `data` is not a `DataFrame` instance...

cuducos replied:
This is better. But a lot is going on, and it is quite a repetitive sequence of checks between lines 73-78 (and you still have two early returns for the same case, empty `data`, one on line 75, another on line 78). Have you considered a function like `is_empty(data)` that returns a boolean to that question?

FelipeSBarros replied:
> This is better. But a lot is going on, and it is quite a repetitive sequence of checks between lines 73-78 (and you still have two early returns for the same case, empty `data`, one on line 75, another on line 78). Have you considered a function like `is_empty(data)` that returns a boolean to that question?

Sure I can. I should be done this before. By the way... Should I be writting tests for `flatten_df`, and `is_empty`, right?

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -66,28 +66,33 @@ def _flatten_df(row, column_name):
     )
 
 
+def is_empty(data):
+    if not HAS_PANDAS or (HAS_PANDAS and not isinstance(data, DataFrame)):
+        return not data
+    elif isinstance(data, DataFrame):
+        return data.empty
+    else:
+        return False
```
Created the `is_empty` function to evaluate the `data` inpput and execute an early return, if necessary.

FelipeSBarros commented on crossfire/__init__.py:
```diff
@@ -66,28 +66,33 @@ def _flatten_df(row, column_name):
-        if not data:
-            return data
+    if is_empty(data):
+        return data
```
Using the created function...

FelipeSBarros commented on tests/test_flatten.py:
```diff
@@ -25,7 +28,11 @@ def teste_flatten_with_emptylist():
 
 
-# test the flatten function with a dictionary mocking it to assert _flatten_dict function is being called
+def teste_flatten_with_emptyDataFrame():
+    data = flatten(DataFrame())
+    assert data.equals(DataFrame())
```
Created this test to assure if an empty `DataFrame` is passed aas `data`, it will be returned buy the early return.
Should I test if `is_empty` function is being called, right?
Just realized this now. I will update it soon...

cuducos replied:
> Created this test to assure if an empty `DataFrame` is passed aas `data`, it will be returned buy the early return.

This test does not assure that. It ensures the result is the “same”, but not “when” it is returned. If you wanna make sure this is happening, you have to mock a method that happens _after_ the early return and [`.assert_not_called()`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_not_called).

FelipeSBarros commented on tests/test_flatten.py:
```diff
@@ -55,3 +61,14 @@ def test_flatten_pd():
     )
+
+
+def test_aux_function_is_called():
+    mock_flatten_df = Mock()
+
+    with pytest.MonkeyPatch().context() as m:
+        m.setattr("crossfire._flatten_df", mock_flatten_df)
+
+        flatten(PD_DATA, nested_columns=["contextInfo"])
+
+    mock_flatten_df.assert_called_once()
```
Mock the `_flatten_df` to assert if it is being called during the `flatten` function recieves a `DataFrame`.
Realizing now that perhaps the name must be improved, right?
Will update it soon

FelipeSBarros reviewed (commented):
OK, now with `ìs_empty` function and some tests...

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +59,44 @@ def occurrences(
+        return not data
+    elif isinstance(data, DataFrame):
+        return data.empty
+    else:
+        return False
```
What is the use case for this `else` branch?

Overall I think the readability is more complex than it could be. In my mind what we need is two main branches of logic:
* If `HAS_PANDAS` we check if data is a `DataFrame` and, if so, return `data.empty`
* If not, just return the boolean representation of `data`

Is there any use case I am leaving behind?

FelipeSBarros replied:
It make sense. I don't now how. But yesterday the way I wrote was making sense...
Mainly the first branch, checking if the user hasn't Pandas...
Any way, I ahve updated using your suggestion.

cuducos commented on tests/test_flatten.py:
```diff
@@ -19,7 +28,11 @@ def teste_flatten_with_emptylist():
 
 
-# test the flatten function with a dictionary mocking it to assert _flatten_dict function is being called
+def teste_flatten_with_emptyDataFrame():
```
Any reason to ignore PEP8 in the [naming](https://peps.python.org/pep-0008/#function-and-variable-names)?

```suggestiotn
def teste_flatten_with_empty_data_frame():
```

FelipeSBarros replied:
Sorry. Mental issues here...
By the way, should I improve the name with the expected result?
like `teste_flatten_with_empty_data_frame_return_empty_data_frame` or `teste_flatten_with_empty_data_frame_yearly_return`?

cuducos replied:
You can, but no need to. The test is short enough to make that clear : )

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +44,31 @@ def test_flatten_dict():
+    )
+
+
+def test_aux_function_is_called():
```
Does the `aux_function` have a name? Do you refer to `is_empty` or `_flatten_df`? 

```suggestion
def test_flatten_df_is_called():
```

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +44,31 @@ def test_flatten_dict():
+
+
+def test_aux_function_is_called():
+    mock_flatten_df = Mock()
+
+    with pytest.MonkeyPatch().context() as m:
+        m.setattr("crossfire._flatten_df", mock_flatten_df)
+
+        flatten(PD_DATA, nested_columns=["contextInfo"])
+
+    mock_flatten_df.assert_called_once()
```
Is there a reason to use the unnecessarily complicated monkey patch from `pytest` instead of the native `unittest.mock` used [in other tests in this project](https://github.com/search?q=repo%3AFelipeSBarros%2Fcrossfire%20unittest.mock&type=code)?

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +61,42 @@ def occurrences(
+def is_empty(data):
+    if HAS_PANDAS and isinstance(data, DataFrame):
+        return data.empty
+    else:
+        return not data
```
What is the role of this `else` if the `if` already returns?

```suggestion
    return not data
```

cuducos commented on crossfire/__init__.py:
```diff
@@ -51,16 +61,42 @@ def occurrences(
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    if HAS_PANDAS and isinstance(data, (DataFrame, GeoDataFrame)):
```
Do we need `GeoDataFrame`?

```suggestion
    if HAS_PANDAS and isinstance(data, DataFrame):
```

Checking the documentation for `isinstance` I think we don't.

cuducos commented on tests/test_flatten.py:
```diff
@@ -8,18 +13,29 @@
 
 
-# test the flatten function with a dictionary mocking it to assert _flatten_dict function is being called
+def teste_flatten_with_empty_data_frame():
```
As far as I know `pytest` runs functions starting with `test_` (not `teste_`). Can you confirm this test is running?

cuducos commented on tests/test_flatten.py:
```diff
@@ -8,18 +13,29 @@
+GEOPD_DATA = GeoDataFrame(DICT_DATA, crs="EPSG:4326", geometry=GEOMETRY)
 
 
 def teste_flatten_wrong_nested_columns_value_error():
```
As far as I know `pytest` runs functions starting with `test_` (not `teste_`). Can you confirm this test is running?

cuducos commented on tests/test_flatten.py:
```diff
@@ -8,18 +13,29 @@
 
-def teste_flatten_with_emptylist():
+def teste_flatten_with_empty_list():
     assert flatten([]) == []
```
Should we test an empty `DataFrame` too?

cuducos commented on tests/test_flatten.py:
```diff
@@ -8,18 +13,29 @@
 
-# test the flatten function with a dictionary mocking it to assert _flatten_dict function is being called
+def teste_flatten_with_empty_data_frame():
+    mock_flatten_df = Mock()
+
+    with patch("crossfire._flatten_df", mock_flatten_df):
```
Please, review the docs for `patch` (or other use cases in this project)  because this test does not do what you think it is doing. It mocks, but the mock you assert is not the patched one (it is the one you created manually on line 31).

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +47,55 @@ def test_flatten_dict():
+
+
+def test_flatten_df_is_called():
+    mock_flatten_df = Mock()
+
+    with patch("crossfire._flatten_df", mock_flatten_df):
```
Please, review the docs for `patch` (or other use cases in this project)  because this test does not do what you think it is doing. It mocks, but the mock you assert is not the patched one (it is the one you created manually on line 70).

cuducos commented on tests/test_flatten.py:
```diff
@@ -67,12 +65,18 @@ def test_flatten_pd():
-    mock_flatten_df = Mock()
-
-    with patch("crossfire._flatten_df", mock_flatten_df):
+    with patch("crossfire._flatten_df") as mock_flatten_df:
+        mock_flatten_df.return_value(
+            Series(
+                {
+                    "contextInfo_contextInfo_context1": "info1",
+                    "contextInfo_contextInfo_context2": "info2",
+                }
+            )
+        )
```
`return_value` is not a function to call.
It is ` mock_flatten_df.return_value = 42` not ` mock_flatten_df.return_value(42)`.

FelipeSBarros reviewed (commented):
I have fixed the mock's ´return_value`. But the test still failing bu not having objects to concatenate...

FelipeSBarros commented:
> I have fixed the mock's ´return_value`. But the test still failing bu not having objects to concatenate...

Reading [this SOF}(https://stackoverflow.com/a/57569598) I saw the suggestion on reading about `autospec=True`...
Will check out this latter...

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +45,60 @@ def test_flatten_dict():
+    with patch("crossfire._flatten_df", autospec=True) as mock_flatten_df:
+        mock_flatten_df.return_value = Series(
+            {
+                "contextInfo_contextInfo_context1": "info1",
+                "contextInfo_contextInfo_context2": "info2",
```
Why does `contextInfo` appear twice here?

Wasn't it supposed to be:

```suggestion
                "contextInfo_context1": "info1",
                "contextInfo_context2": "info2",
```

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +45,60 @@ def test_flatten_dict():
+
+
+def test_flatten_df_is_called():
+    with patch("crossfire._flatten_df", autospec=True) as mock_flatten_df:
```
What problem does the `autospec` fixes here? You shared that tests were not passing, but you haven't commented on _why_ they weren't — what was expected? Was there an error? 

I am just trying to understand why we need `autospec` because I have the feeling we don't (e.g. I think there's an error in the return value above).

FelipeSBarros replied:
You are rght. I can see that [my comment](https://github.com/FelipeSBarros/crossfire/pull/103#pullrequestreview-1909224735) could be more soecific.

So, basically, even with the `return_value` set correctly, the `contact` function, which is executed in sequence, [fails due inexistence of objects to contatenate](https://github.com/FelipeSBarros/crossfire/actions/runs/8098685427/job/22132667120#step:6:107):

> FAILED tests/test_flatten.py::test_flatten_df_is_called - ValueError: No objects to concatenate

I still not having sure about what the `sutospec` does. So, not sure, what it is fixing...

cuducos replied:
>  I still not having sure about what the sutospec does.

Ok, so **never** add code you don't understand. Never. You can use it to ask: hey, I noticed that when I add that, it works for you. But this is a way to explore the issue rather than a solution.

> No objects to concatenate

So my best guess is that something unexpected happens [when you call the `concat` function](https://github.com/FelipeSBarros/crossfire/blob/f7a01447cd7b722691da7572def59dbc64045d73/crossfire/__init__.py#L84-L90). Have you tried to debug and understand the problem? Try to describe what is happening and why this specific error is popping up.

One idea is to use `import pytest; pytest.set_trace()` just before `concat` and inspect every look of the `for` loop.

FelipeSBarros replied:
> Ok, so **never** add code you don't understand. Never. You can use it to ask: hey, I noticed that when I add that, it works for you. But this is a way to explore the issue rather than a solution.

Thanks for your advice!

> So my best guess is that something unexpected happens [when you call the `concat` function](https://github.com/FelipeSBarros/crossfire/blob/f7a01447cd7b722691da7572def59dbc64045d73/crossfire/__init__.py#L84-L90). Have you tried to debug and understand the problem? Try to describe what is happening and why this specific error is popping up.
> 
> One idea is to use `import pytest; pytest.set_trace()` just before `concat` and inspect every look of the `for` loop.

OK, So I will work os that.

FelipeSBarros replied:
@cuducos I have been trying to understand the issue with the `test_flatten_df_is_called`, using `pytest.set_trace()`.
using the mock with parameter `autospec` set to `False` (as default). 
The execution of `data.apply(_flatten_df, args=(key,), axis=1)` return the error: "ValueError: No objects to concatenate"

```
poetry run pytest -k test_flatten_df_is_called
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.2, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/felipe/repos/crossfire
configfile: pyproject.toml
plugins: ruff-0.2.1, anyio-4.1.0, asyncio-0.21.1
asyncio: mode=strict
collected 94 items / 93 deselected / 1 selected                                                                                                                                          

tests/test_flatten.py 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/felipe/repos/crossfire/crossfire/__init__.py(86)flatten()
-> data = concat(
(Pdb) data.apply(_flatten_df, args=(key,), axis=1)
*** ValueError: No objects to concatenate

# confirming mock 
(Pdb) _flatten_df
<MagicMock name='_flatten_df' id='139870010380432'>
(Pdb) _flatten_df.return_value
contextInfo_context1    info1
contextInfo_context2    info2
dtype: object
```

But running the function without mocking it it returns no ValueError.
I tried to uderstand the issue using chatgpt without sucess until i found [this issues](https://github.com/pandas-dev/pandas/issues/45298) on github.
So, seems to be a problem in the way the mocked function is being used:
> In a context manager, a mocked function in Series.apply calls the function to aggregate instead of the standard one.

Probably, this is why the mock I did is returning ValueErrror related to `concatenate`, as it is being used to `aggregate` instead. It is interesting becuse ocasionally we are using a concatenate function just after its evaluation.

Any way, According to [this comment](https://github.com/pandas-dev/pandas/issues/45298#issuecomment-1009768562) we could overcome the issue by setting `nre_valleble=Mock` in the context manager.
I have tested here and it seems to work well.
```python
def test_flatten_df_is_called():
    with patch("crossfire._flatten_df", new_callable=Mock) as mock_flatten_df:
        mock_flatten_df.return_value = Series(
            {
                "contextInfo_context1": "info1",
                "contextInfo_context2": "info2",
            }
        )

        flatten(PD_DATA, nested_columns=["contextInfo"])

    mock_flatten_df.assert_called_once()
```

But I dont feel confident to commit this code as I couldn't garantee that this was the real issue in the code.

Need your help... :)

cuducos replied:
In cases like that, I like to document the _why_ because the code does not tell the whole story, e.g.:

```python
    # There is a bug on Pandas that makes contact fails with the default MagicMock
    with patch("crossfire._flatten_df", new_callable=Mock) as mock_flatten_df:
```

cuducos commented on crossfire/__init__.py:
Do you think these functions belong in the main module? If so, can you expand on the _why_.

FelipeSBarros replied:
I have already though about it...
As we are expecting to recieve `Occurrences`, perharps it should be in the `Occurrences` module.
What is your opinion? Is it is the case of moving to `Occurrences`, would be OK to finish it first and then move it to the right place?

FelipeSBarros reviewed (commented):
@cuducos I have changed the `flatten` function to `crossfire.clients.occurrences`.

cuducos commented on crossfire/clients/occurrences.py:
```diff
@@ -164,3 +177,43 @@ def __call__(self):
+    if is_empty(data):
+        return data
+    if HAS_PANDAS and isinstance(data, DataFrame):
+        keys = set(data.columns) & nested_columns
+        for key in keys:
+            data = concat(
+                (
+                    data.drop(key, axis=1),
+                    data.apply(_flatten_df, args=(key,), axis=1),
+                ),
+                axis=1,
+            )
```
Don't you think all this logic should be `_flatten_df`?

You can have the auxiliary function inside it, e.g.:

```python
def _flatten_df(data):
    def _flatten_col(col_name):
        # ...
    # code like data.apply(_flatten_col, col_name)
```

cuducos commented on crossfire/clients/occurrences.py:
```diff
@@ -164,3 +177,43 @@ def __call__(self):
+
+        return data
+
+    keys = set(data[0].keys()) & nested_columns
+    for item in data:
+        for key in keys:
+            item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
+            item.pop(key)
```
What about extracting this as a `_flatten_list` to be a parallel of `_flatten_df`?

FelipeSBarros replied:
@cuducos in the last commit I did this extraction. Hope the way I have done is fine.

cuducos commented on tests/test_flatten.py:
```diff
@@ -31,3 +46,63 @@ def test_flatten_dict():
+
+
+def test_flatten_df_is_called():
+    # There is a bug on Pandas that makes apply fails when called from Series with the default MagicMock
```
Can you add the link to the issue (bug report) as a comment too?

FelipeSBarros replied:
Sure, I will. Also I have subcribed to the issue to get news about it.

FelipeSBarros commented:
@cuducos, sorry for the long time without working on it.

> Something I have asked before and don't recall a clear answer: have you run this code (`pytest`) without Pandas installed?

I have decided to face this wuestion by installing the pakcge without extras and running `pytest`...
The following error were found (obviously):
```python
poetry run pytest
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.2, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/felipe/repos/crossfire
configfile: pyproject.toml
plugins: ruff-0.2.1, anyio-4.1.0, asyncio-0.21.1
asyncio: mode=strict
collected 58 items / 3 errors                                                                                                                                                            

========================================================================================= ERRORS =========================================================================================
_________________________________________________________________________ ERROR collecting tests/test_flatten.py _________________________________________________________________________
ImportError while importing test module '/home/felipe/repos/crossfire/tests/test_flatten.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.10.2/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_flatten.py:3: in <module>
    from geopandas import GeoDataFrame
E   ModuleNotFoundError: No module named 'geopandas'
___________________________________________________________________ ERROR collecting tests/test_occurrences_client.py ____________________________________________________________________
ImportError while importing test module '/home/felipe/repos/crossfire/tests/test_occurrences_client.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.10.2/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_occurrences_client.py:3: in <module>
    from geopandas import GeoDataFrame
E   ModuleNotFoundError: No module named 'geopandas'
_________________________________________________________________________ ERROR collecting tests/test_parser.py __________________________________________________________________________
ImportError while importing test module '/home/felipe/repos/crossfire/tests/test_parser.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.pyenv/versions/3.10.2/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_parser.py:4: in <module>
    from geopandas import GeoDataFrame
E   ModuleNotFoundError: No module named 'geopandas'
================================================================================ short test summary info =================================================================================
ERROR tests/test_flatten.py
ERROR tests/test_occurrences_client.py
ERROR tests/test_parser.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=================================================================================== 3 errors in 0.16s ====================================================================================

```
So, I started addin a try/except in the import section, just like we did in [occurrences](https://github.com/FelipeSBarros/crossfire/blob/master/crossfire/clients/occurrences.py#L9-L12)
```python
try:
    from pandas import concat
except ImportError:
    pass
```

By doing something similar to the above code, I could solve the `ModuleNotFoundError:` error found before.
But, after that all test that uses `pandas` or `geopandas` packages failed by missing clases, method, etc from those packages.

I asked ChatGPT for a few suggestions on how to solve this issue, which I share here, before implementing as I saw from a few posts from you whowing that it could give inhapropriate answers:
Its suggestion was to adapt `mark` function from pytest to create a `skip_if_pandas_not_installed` marker, indicating whether a test must be skipped if `´pandas`, `geopandas` o `shapely` is not installed:
`skip_if_pandas_not_installed = mark.skipif(not HAS_PANDAS, reason="pandas is not installed")`

What do you think about this suggestion?

cuducos commented:
I like it :)

FelipeSBarros commented on pyproject.toml:
```diff
@@ -20,7 +20,6 @@ exclude =  ["env-example"]
 python = "^3.9,<3.13"
 geopandas = { version = "^0.13.2", optional = true }
 httpx = "^0.25.0"
-pandas = { version = "^2.1.1", optional = true }
```
This mistake I will resolve in the next commit...

cuducos replied:
Is there a reason not to add it back where it was? Packages were ordered alphabetically, but now they are not.

cuducos commented on .github/workflows/github-actions.yml:
```diff
@@ -17,9 +17,15 @@ jobs:
       - name: Install and configure Poetry
         uses: snok/install-poetry@v1
-      - name: Install dependencies
+      - name: Install all dependencies
         run: |
           poetry install --all-extras
       - name: Run tests
         run: |
           poetry run pytest
+      - name: Install basic dependencies
+        run: |
+          poetry install
```
Why install all dependencies, too? Later, install “basic” dependencies (which are already included in the “all” ones)?

FelipeSBarros replied:
Perhaps I should find a better name and invert thte execution process....
So, first install the "basic" dependencies dosen't install pandas and geopandas, to run the test without those packages.

Then, install the dependencies with extras, which means, pandas and geopandas...
What do yout think?

cuducos replied:
> invert thte execution process....

I think you need two separate jobs/processes on the CI. Maybe three. One without Pandas or GeoPandas, one with Pandas (but not GeoPandas), and another one with GeoPandas.

FelipeSBarros replied:
>I think you need two separate and independent jobs/processes on the CI. Maybe three. One without Pandas or GeoPandas, one with Pandas (but not GeoPandas), and another one with GeoPandas.
I agree. I will revert the change I have done and add those jobs on a different PR.
ok?

cuducos replied:
Have you manually tested executing the library without Pandas and without GeoPandas? Otehrwise we need it in this PR.

FelipeSBarros replied:
Yes I have tested manually without pandas and geopandas, with pandas only, and with pandas and geopandas.

cuducos replied:
Ok. But keeping [the logic that arguably does not make sense](https://github.com/FelipeSBarros/crossfire/pull/103/files#r1531416332)?

FelipeSBarros replied:
@cuducos you made me thought about the importance of having this issue solved in this PR.
Because of that, Ihave decided to commit an proposal that try yo solve the #104 ;

Reading about [github flow](https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow) I saw that e can have as many jobs as it is needed.
In our case, I note at least three is needed:
    1. one to test several python versions with the basic dependencies (i.e. without Pandas and GeoPandas); 
    2. another to test several python versions with Pandas dependencies only and, 
    3. the last one to test several python versions with both Pandas and GeoPandas dependencies.

Trying to make it more organized, I set the prefix `(basic dependencies)`, `(Pandas only)` and `(Pandas & GeoPandas)` before the Python version information and, also, I set [prerequisites](https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#example-requiring-successful-dependent-jobs) between jobs, so we first have the test run with basic dependencies; once finished, it will run the test with Pandas only and, then, if passing, the test with both Pandas and GeoPandas.

I understand that perhaps this is a misusing of this feature, and that making one set of tests require onthers to pass make it take longer. But I like the way it keeps thinks more organized...

![image](https://github.com/FelipeSBarros/crossfire/assets/7208674/36f413a0-d4ab-495c-b474-aeecefb7915b)

cuducos commented on tests/test_client.py:
```diff
@@ -14,6 +15,14 @@
 )
 from crossfire.parser import UnknownFormatError
 
+if importlib.util.find_spec("pandas"):
+    HAS_PANDAS = True
+else:
+    HAS_PANDAS = False
```
Is there any reason to adopt a different [pattern than the one already used in the package](https://github.com/search?q=repo%3AFelipeSBarros%2Fcrossfire+HAS_PANDAS&type=code)?

cuducos commented on tests/test_flatten.py:
```diff
@@ -1,25 +1,65 @@
 from pytest import raises
 
-from crossfire import NestedColumnError, flatten
+try:
+    from shapely import Point
+except ImportError:
+    pass
```
Why do we need if `shapely` is not in our requirements?

FelipeSBarros replied:
ctuakky it is a dependency from Geopandas. So, it is not in the "pyproject.toml". I think I can add this together in the ty/excepto gepándas importing:
```
try:
    from geopandas import GeoDataFrame
    from shapely import Point

    HAS_GEOPANDAS = True
except ModuleNotFoundError:
    HAS_GEOPANDAS = False
```
This way is better?

FelipeSBarros commented on crossfire/clients/occurrences.py:
```diff
@@ -181,11 +181,26 @@ def __call__(self):
-    return Series(
-        {f"{column_name}_{key}": value for key, value in column_data.items()}
-    )
+def _flatten_df(data, nested_columns):
+    def _flatten_col(row, column_name):
+        column_data = row[column_name]
+        return Series(
+            {
+                f"{column_name}_{key}": value
+                for key, value in column_data.items()
+            }
+        )
+
+    keys = set(data.columns) & nested_columns
+    for key in keys:
+        data = concat(
+            [
+                data.drop(key, axis=1),
+                data.apply(_flatten_col, args=(key,), axis=1),
+            ],
+            axis=1,
+        )
+    return data
```
Commit created to addres [your comment](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1511572073)..

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -1,8 +1,21 @@
+
+    HAS_GEOPANDAS = True
+except ImportError:
+    HAS_GEOPANDAS = False
+try:
```
Readability (if autoformat allows)

```suggestion
    HAS_GEOPANDAS = False

try:
```

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -1,8 +1,21 @@
+
+    HAS_PANDAS = True
+except ImportError:
+    HAS_PANDAS = False
+try:
```
Readability (if autoformat allows)

```suggestion
    HAS_PANDAS = False

try:
```

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -1,8 +1,21 @@
+try:
+    from pandas.testing import assert_frame_equal
+except ImportError:
+    pass
 from pytest import mark, raises
```
Readability (if autoformat allows)

```suggestion
    pass

from pytest import mark, raises
```

FelipeSBarros commented:
I thought on two way of working araound in [`_flatten_list`](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/crossfire/clients/occurrences.py#L206):

https://github.com/FelipeSBarros/crossfire/blob/f0a01577f95cfd0470e3e4ff3eb1b98157c36a30/crossfire/clients/occurrences.py#L208-L210

1. using a validation to confirm if a `item.get(key)` is not `None`;
2. or simply adding an `if v` in the dict comprehension;

cuducos commented:
> 1. using a validation to confirm if a `item.get(key)` is not `None`;
> 2. or simply adding an `if v` in the dict comprehension;

I think both of them might discard data that is important. `if v` will fail for `0` (which is different than `None`) and well as `if item.get(key)` fails for `item[key] = 0`. I think what you need is to know:

1. if the key is in the item
1. if ~~key is~~ in ~~the~~ item
1. if key in item
1. `if key in item`

FelipeSBarros commented:
> > 1. using a validation to confirm if a `item.get(key)` is not `None`;
> > 2. or simply adding an `if v` in the dict comprehension;
> 
> I think both of them might discard data that is important. `if v` will fail for `0` (which is different than `None`) and well as `if item.get(key)` fails for `item[key] = 0`. I think what you need is to know:
> 
> 1. if the key is in the item
> 2. if ~key is~ in ~the~ item
> 3. if key in item
> 4. `if key in item`

@cuducos first of all sorry for the belated answer. I must confess that I couldn't understand your last message.
So, today I could dedicate a spare time on trying a solution to make the test pass. I just worked on it. After having all test passing I will work on your proposal of using the `flatten` as an `Occurrence` parameter, and not as a function it self.
Let me know if a could get your point on the last message or if I still work to do.

cuducos commented on crossfire/clients/occurrences.py:
```diff
@@ -164,3 +179,58 @@ def __call__(self):
+    keys = set(data[0].keys()) & nested_columns
+    for item in data:
+        for key in keys:
+            if key in item.keys():
+                if item.get(key):
+                    item.update(
+                        {f"{key}_{k}": v for k, v in item.get(key).items() if v}
+                    )
+                    item.pop(key)
```
These lines have four lookups for the keys in the dictionary. This is unnecessary.

But first, why `.keys()` if the lookup is already in the keys?

```suggestion
            if key in item:
                if item.get(key):
                    item.update(
                        {f"{key}_{k}": v for k, v in item.get(key).items() if v}
                    )
                    item.pop(key)
```

Now, let's save the result of the first lookup and re-use instead of performing another lookup:

```suggestion
            if key in item:
                value = item.get(key)
                if value:
                    item.update(
                        {f"{key}_{k}": v for k, v in value.items() if v}
                    )
                    item.pop(key)
```

Now, let's get the result of the `.pop` instead of another lookup to remove the item:

```suggestion
            if key in item:
                value = item.pop(key)
                if value:
                    item.update(
                        {f"{key}_{k}": v for k, v in value.items() if v}
                    )
```

FelipeSBarros commented:
@cuducos thanks for your corrections. It is much better and I could see that the test was with a wrong expected return value.
I wasn't considering that if a key has no value it should be removed anyway, which makes sense on a list of dictionary with mixed values: some "contextInfo" with nested values and some without, for instance...
```python
data = [
    {
        "answer": 42,
        "contextInfo": {'context1': 'info1',
    },
    {
        "answer": 42,
        "contextInfo": None,
    },
]
```
So in this case the `_flatten_list`would return:

```python
[{'answer': 42, 'contextInfo_context1': 'info1'}, {'answer': 42}]
```
Now, if I have a list of dict with all keys without nested data, it will drop the key we are trying to flatten. For instance:
```python
data = [
    {
        "answer": 42,
        "contextInfo": None,
    },
    {
        "answer": 42,
        "contextInfo": None,
    },
]
# _flatten_list function Would return
[{'answer': 42}, {'answer': 42}]
```
So, in this situation it is not flattening anything (as there is nothig to be flatten) but is doing something that (I believe) wouldn't be expected: it is dropping the keys. Make sense my concerns?

cuducos commented on README.md:
````diff
@@ -125,6 +125,40 @@ occurrences('813ca36b-91e3-4a18-b408-60b27a1942ef', format='geodf')
+
+```python
+from crossfire.clients.occurrences import flatten
+flatten(data, nested_columns=None)
````
If it is optional, do we need to pass `None`?

```suggestion
flatten(data)
```

FelipeSBarros commented:
@cuducos adicionei um teste para começar a trabalhar na situação [mencionada anteriormente](https://github.com/FelipeSBarros/crossfire/pull/103#issuecomment-2102860151). Didn't worked making the test pass because I can see two ways of doing it.
The first one is, not "popping" the key under evaluation. Thus it would change the return value from all testes, independently of it having or not a nested keys and values on it;
The second one, which I don't like much , is adding an "else" statement for `if value:`, so we could update the dict inserting back the key with None value. This last option seems to be the worst one.
this last options seems a not good approach as we are adding complexity to the code and to the result. Some dicts would have the flattened key and values, while others only the orignal key with None as value.

Would you suggest any other apporach?

cuducos commented:
I agree with the last approach, but… why popping the key just to reinsert it next? Can’t you prevent popping it inste?

FelipeSBarros commented:
> I agree with the last approach, but… why popping the key just to reinsert it next? Can’t you prevent popping it instead?

@cuducos , I am not sure if I could understand your last comment. You agree it is the worst scenarious? I see the first proposal as the most insteresting: not "popping" anythink. just flattenning when there is something to be flattened.
This way the original key will be always kept. And those dicts with nested values would have new keys with the nested values flatened.

cuducos commented on crossfire/clients/occurrences.py:
```diff
@@ -164,3 +179,58 @@ def __call__(self):
+    keys = set(data[0].keys()) & nested_columns
+    for item in data:
+        for key in keys:
+            if key in item:
+                value = item.get(key)
+                if value:
+                    item.update(
+                        {f"{key}_{k}": v for k, v in value.items() if v}
+                    )
```
Try to keep the [flat path to success](https://www.vinta.com.br/blog/flat-success-path)


```suggestion
            if key not in item:
                continue
                
            value = item.get(key)
            if not value:
                continue

            item.update({f"{key}_{k}": v for k, v in value.items() if v})
```

FelipeSBarros commented:
@cuducos in my last commits I have refactored the `_flaten_df` so it has the same behavior as `_flatten_list`, when dealing with DataFrame with None values in the potential `nested_columns`;
Also, before merging this PR, I decided to follow your [suggestion](https://github.com/FelipeSBarros/crossfire/issues/96#issuecomment-2068152514): expose `flatten` functionality, not as a function.

I added the `flat` evaluation in `Occurrences`' [`__call__`](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/crossfire/clients/occurrences.py#L137-L150)
 method, and updated the tests.

I am missing a integration test to confirm if when I call `Occurrences(..., flat=True)`, the `flatten()` function is being called. but I am not sure if it is necessary.

FelipeSBarros commented:
few things I still have to do:
1. Confirm/fix if implementation of `flat` paramter is fine;

Trying the implementation I realized that when requesting data as ´dict` format, the `flaten` function is not being used:
```python
from crossfire import occurrences

occs_dict = occurrences(id_state='813ca36b-91e3-4a18-b408-60b27a1942ef',
                id_cities='5bd3bfe5-4989-4bc3-a646-fe77a876fce0',
                initial_date='2018-04-01', flat=True, format='dict')
'contextInfo_mainReason' in occs_dict[0].keys()
#False

occs_df = occurrences(id_state='813ca36b-91e3-4a18-b408-60b27a1942ef',
                id_cities='5bd3bfe5-4989-4bc3-a646-fe77a876fce0',
                initial_date='2018-04-01', flat=True, format='df')
'contextInfo_mainReason' in occs_df.columns
#True

occs_gdf = occurrences(id_state='813ca36b-91e3-4a18-b408-60b27a1942ef',
                id_cities='5bd3bfe5-4989-4bc3-a646-fe77a876fce0',
                initial_date='2018-04-01', flat=True, format='geodf')
'contextInfo_mainReason' in occs_gdf.columns
#True
```

2. Figure out how to deal with nested columns in nested columns;

When trying the implementation I realized that `contextInfo` usually come with other nested key/columns as values:

```python
occs_df = occurrences(id_state='813ca36b-91e3-4a18-b408-60b27a1942ef',
                id_cities='5bd3bfe5-4989-4bc3-a646-fe77a876fce0',
                initial_date='2018-04-01', flat=True, format='df')
'contextInfo_mainReason' in occs_df.columns

occs_df.iloc[0]['contextInfo_mainReason']
#{'id': 'baa3b299-67ad-41d2-aaf0-23ec8288cadb', 'name': 'Homicidio/Tentativa'}
```
How could we confirm if flattened key/column come with nested value and flatten then also?

3. Rewrite documentation;

cuducos commented:
> How could we confirm if flattened key/column come with nested value and flatten then also?
I think the point is checking the _type_ of value:

```python
for key, value in occurrence.items():
    if isinstance(value, dict):
        flat(value)
```

Maybe for Pandas it would be a string formated as JSON, so you can try to parse it as JSON and check if it is a dictionary:

```python
def is_nested(text):
    try:
        data = loads(text)
    except ValueError:  # not sure this is the right exception, this is just an example
        return False
    return isinstance(data, dict)
```

FelipeSBarros commented:
> > How could we confirm if flattened key/column come with nested value and flatten then also?
> > I think the point is to check the _type_ of value:
> 
> ```python
> for key, value in occurrence.items():
>     if isinstance(value, dict):
>         flat(value)
> ```
> 
> Maybe for Pandas it would be a string formatted as JSON, so you can try to parse it as JSON and check if it is a dictionary:
> 
> ```python
> def is_nested(text):
>     try:
>         data = loads(text)
>     except ValueError:  # not sure this is the right exception, this is just an example
>         return False
>     return isinstance(data, dict)
> ```

Thanks for your advice. I hope I have implemented the right way...

By the way, I am facing a new problem:
when I run the test [`test_flatten_pd_with_nested_columns_with_nested_values`](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/tests/test_flatten.py#L263) only, it pass:

```python
poetry run pytest -k test_flatten_pd_with_nested_columns_with_nested_values
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.2, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/felipe/repos/crossfire
configfile: pyproject.toml
plugins: ruff-0.2.1, anyio-4.1.0, asyncio-0.21.1
asyncio: mode=strict
collected 104 items / 103 deselected / 1 selected                                                                                                                                        

tests/test_flatten.py .                                                                                                                                                            [100%]

```
But running all tests, it fails.
```command
FAILED tests/test_flatten.py::test_flatten_pd_with_nested_columns_with_nested_values - AssertionError: DataFrame are different

```
I have no idea of what might be causing the probem... do you have any clue?

cuducos commented:
> when I run the test [test_flatten_pd_with_nested_columns_with_nested_values](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/tests/test_flatten.py#L263) only, it pass:
>
> [...]                                                                                                                                                         [100%]
>
> But running all tests, it fails.

Probably because when you run [this](https://github.com/FelipeSBarros/crossfire/blob/4270969134766c91e19c998f65cd251a540132fc/tests/test_flatten.py#L254-L257) you mutate `DICT_DATA_WITH_NESTED_VALUES_IN_NESTED_COLUMNS` — `flatten` does not return a new dictionary, it mutates the one passed as an argument. So the input is different on the second test when you run both tests. You can copy the dictionary in each test to prevent that.

FelipeSBarros commented:
> > when I run the test [test_flatten_pd_with_nested_columns_with_nested_values](https://github.com/FelipeSBarros/crossfire/blob/make_flatten_gdf/tests/test_flatten.py#L263) only, it pass:
> > [...]                                                                                                                                                         [100%]
> > But running all tests, it fails.
> 
> Probably because when you run [this](https://github.com/FelipeSBarros/crossfire/blob/4270969134766c91e19c998f65cd251a540132fc/tests/test_flatten.py#L254-L257) you mutate `DICT_DATA_WITH_NESTED_VALUES_IN_NESTED_COLUMNS` — `flatten` does not return a new dictionary, it mutates the one passed as an argument. So the input is different on the second test when you run both tests. You can copy the dictionary in each test to prevent that.

Thanks, Cuducos! I had this in mind last night. But I was so tired that I didn't realize that I should use a [deepcopy](https://docs.python.org/3/library/copy.html#copy.deepcopy) to make sure all values has copied not ony the structure.


Could you review the implementation of `flatten()` function as an Occurrence parameter (point 1 [here](https://github.com/FelipeSBarros/crossfire/pull/103#issuecomment-2112777531) )? I am not confident about the way I have implemented... I also it is not being executed when requesting the `Occurrences ` as `dict`. :/

cuducos commented on tests/test_flatten.py:
```diff
@@ -1,33 +1,274 @@
-from pytest import raises
+from copy import deepcopy
```
Why `deepcopy` instead of [`dict.copy`](https://docs.python.org/3/library/stdtypes.html#dict.copy)?

FelipeSBarros replied:
Because copy, as it is mentioned int he documentation:
> Return a shallow copy of the dictionary.
So, even using it (copy()) the test fail because of the use of `DICT_DATA_WITH_NESTED_VALUES_IN_NESTED_COLUMNS` in two differents tests that alter its structure.
I understood, it copies the data strucutre.
I understand that `copy()` creates a new object but keeps references with the original. And because of that, changes made in the copy affects thte original.

The [`deepcopy`](https://python.readthedocs.io/en/v2.7.2/library/copy.html#copy.deepcopy), however, creates a full copy of the data:
> A deep copy constructs a new compound object and then, recursively, inserts copies into it of the objects found in the original.

cuducos replied:
Nice!

cuducos reviewed (changes requested):
> it is not being executed when requesting the Occurrences as dict. :/

So, the test suite has a bug? It is green, meaning `test_occurrences_as_list_dicts_with_flat_parameter` is passing — how is it passing if `flatten` is not being called? Or does this test has a deceiving name?

FelipeSBarros commented:
> So, the test suite has a bug? It is green, meaning `test_occurrences_as_list_dicts_with_flat_parameter` is passing — how is it passing if `flatten` is not being called? Or does this test has a deceiving name?

Yes, it seems I am the problem... I have just confirmed that the mentioned test is well written and passes. I also did a manual test requesting data from the API using `format='dict'` and it came as expected...
Let's move on.

FelipeSBarros commented:
@cuducos I have updated the REAMDE adding `flat` as an `Occurrences`' parameter. I also added a small section to explaining its meaning and an example of the data returned by using it. Let me know if you agree.

cuducos reviewed (approved):
Really great work : )

FelipeSBarros commented:
> Really great work : )

I am glad to hear it that! Thank you for your patience and teaching.
